### PR TITLE
Override locality load balancer settings per host

### LINKED ISSUE #17802: Support Locality Load Balancer settings override per host
**Describe the feature request**
We have a use case to support where some services run active active (with uneven distribution of traffic) while most others run in a failover mode. Currently, locality load balancing feature is supported at the global mesh config level. There isn't a way to either disable or override there locality load balancing settings per host. Hence, the use case above in't configurable. I see that it was discussed before [here](https://docs.google.com/document/d/16DvMkaLYPam9JWUH_QStO_2gxZPeXIcL067gEbRWNSU/edit) and wasn't pursued. 

**Describe alternatives you've considered**
None, there isn't an alternative

**Affected product area (please put an X in all that apply)**

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[X] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Additional context**
Can this be supported by adding an ability to override locality load balancer settings at the Destination rule. I see that outlier detection on destination rule is already used to control this feature, we could as well extend it to override the global behavior.

CC @jwebb49

### PR DESCRIPTION
**Pull request by aattuluri**
Fixes https://github.com/istio/istio/issues/17802

### CODE REVIEW

**Review by howardjohn — COMMENTED** [review]
Nice! Can you change `tests/integration/pilot/locality` to use a destination rule? Basically should just need to remove the `setupConfig` function and then modify the DR to add the locality settings there. There is already a DR to add the outlier detection so can just add to that one

File: tests/integration/pilot/locality/main_test.go [hunk 1]
```diff
@@ -96,6 +96,11 @@ metadata:
 spec:
   host: {{.Host}}
   trafficPolicy:
+	loadBalancer:
+      localityLbSetting:
```
Comment (howardjohn) [inline tests/integration/pilot/locality/main_test.go]: indent looks off

Comment (howardjohn) [reply tests/integration/pilot/locality/main_test.go]: You can replace L133 with
```
		SetupOnEnv(environment.Kube, istio.Setup(&ist, nil)).
```

**Comment by jasonwzm** [comment]
@aattuluri you will need to use the generated CRDs you updated in https://github.com/istio/api/blob/master/kubernetes/customresourcedefinitions.gen.yaml. Please rebase and use the `make update-crds` command to update that.

**Comment by aattuluri** [comment]
> @aattuluri you will need to use the generated CRDs you updated in https://github.com/istio/api/blob/master/kubernetes/customresourcedefinitions.gen.yaml. Please rebase and use the `make update-crds` command to update that.

Thank you.

**Comment by jasonwzm** [comment]
@aattuluri it looks like the testdata is still not valid.

**Comment by aattuluri** [comment]
> @aattuluri it looks like the testdata is still not valid.

Yes, I will work on that, looks like the api might need to change.

**Comment by aattuluri** [comment]
@howardjohn I think that the checks failed (`gencheck_istio` and `unit-tests_istio`) are not related to my change.

Have a question, the `loadBalancer` needs `lbPolicy` to be specified, else its failing validation in galley.

**Bottom line:** `localityLbSettings` cannot be configured on destination rule unless `loadBalancer` has a `lbPolicy` is specified.

I want to make `lbPolicy` optional, haven't had much success so far.

File: pilot/pkg/networking/core/v1alpha3/cluster.go [hunk 2]
```diff
@@ -138,7 +136,6 @@ func (configgen *ConfigGeneratorImpl) BuildClusters(env *model.Environment, prox
 			outboundClusters = append(outboundClusters, configgen.buildOutboundSniDnatClusters(env, proxy, push)...)
 		}
 		// apply load balancer setting for cluster endpoints
-		applyLocalityLBSetting(proxy.Locality, outboundClusters, env.Mesh.LocalityLbSetting)
```
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/cluster.go]: How do you apply to sni-dnat clusters?

Comment (hzxuzhonghu) [reply pilot/pkg/networking/core/v1alpha3/cluster.go]: Oh, ic. You moved it to applyTrafficPolicy

File: pilot/pkg/networking/core/v1alpha3/cluster.go [hunk 3]
```diff
@@ -840,13 +837,20 @@ func applyTrafficPolicy(opts buildClusterOpts, proxy *model.Proxy) {
 	applyConnectionPool(opts.env, opts.cluster, connectionPool, opts.direction)
 	applyOutlierDetection(opts.cluster, outlierDetection)
 	applyLoadBalancer(opts.cluster, loadBalancer, opts.port, proxy)
+	// Use locality lb settings from traffic policy if present, else use mesh wide locality lb settings
+	var localityLbSettings = opts.env.Mesh.LocalityLbSetting
+	if loadBalancer != nil && loadBalancer.LocalityLbSetting != nil {
+		localityLbSettings = loadBalancer.LocalityLbSetting
+	}
+	applyLocalityLBSetting(proxy.Locality, opts.cluster, localityLbSettings)
```
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/cluster.go]: How about moving this into  `applyLoadBalancer`?

File: pilot/pkg/networking/core/v1alpha3/cluster.go [hunk 4]
```diff
@@ -840,13 +837,20 @@ func applyTrafficPolicy(opts buildClusterOpts, proxy *model.Proxy) {
...
 	if opts.clusterMode != SniDnatClusterMode && opts.direction != model.TrafficDirectionInbound {
 		autoMTLSEnabled := opts.env.Mesh.GetEnableAutoMtls().Value
 		var mtlsCtxType mtlsContextType
 		tls, mtlsCtxType = conditionallyConvertToIstioMtls(tls, opts.serviceAccounts, opts.sni, opts.proxy,
 			autoMTLSEnabled, opts.meshExternal, opts.serviceMTLSMode)
 		applyUpstreamTLSSettings(opts.env, opts.cluster, tls, mtlsCtxType, opts.proxy)
 	}
+
```
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/cluster.go]: nit: remove

File: pilot/pkg/proxy/envoy/v2/eds.go [hunk 5]
```diff
@@ -739,28 +743,31 @@ func getDestinationRule(push *model.PushContext, proxy *model.Proxy, hostname ho
...
-		return false
+		return false, nil
 	}
 
-	_, outlierDetection, _, _ := networking.SelectTrafficPolicyComponents(destinationRule.TrafficPolicy, port)
+	_, outlierDetection, loadBalancerSettings, _ := networking.SelectTrafficPolicyComponents(destinationRule.TrafficPolicy, port)
+	lbSettings = loadBalancerSettings
 	if outlierDetection != nil {
-		return true
+		return true, loadBalancerSettings
```
Comment (hzxuzhonghu) [inline pilot/pkg/proxy/envoy/v2/eds.go]: This is not right. The loadBalancerSettings orders should be subset.TrafficPolicy --> destinationrule.TrafficPolicy --> global mesh config. But here, you reversed the first two orders.

**Review by hzxuzhonghu — CHANGES_REQUESTED** [review]

File: pilot/pkg/proxy/envoy/v2/eds.go (hunk 5 above)
Comment (aattuluri) [reply pilot/pkg/proxy/envoy/v2/eds.go]: Good catch, updated.

File: pilot/pkg/networking/core/v1alpha3/cluster.go [hunk 6]
```diff
@@ -987,16 +986,7 @@ func applyOutlierDetection(cluster *apiv2.Cluster, outlier *networking.OutlierDe
 	}
 }
 
-func applyLoadBalancer(cluster *apiv2.Cluster, lb *networking.LoadBalancerSettings, port *model.Port, proxy *model.Proxy) {
-	if cluster.OutlierDetection != nil {
```
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/cluster.go]: This changed the intended behavior: if no `LoadBalancerSettings` specified, the locality setting of mesh will not be applied.

Comment (hzxuzhonghu) [reply pilot/pkg/networking/core/v1alpha3/cluster.go]: So maybe you should set mesh locality settings to lb in `SelectTrafficPolicyComponents`

Comment (aattuluri) [reply pilot/pkg/networking/core/v1alpha3/cluster.go]: Updated.

**Comment by rshriram** [comment]
> localityLbSettings cannot be configured on destination rule unless loadBalancer has a lbPolicy specified.

It’s just one enum. Why is that an issue ?

**Comment by aattuluri** [comment]
> > localityLbSettings cannot be configured on destination rule unless loadBalancer has a lbPolicy specified.
> 
> It’s just one enum. Why is that an issue ?

It's not an issue, it's more of a usability concern, anyone overriding `localityLbSettings` in destination rule also needs a `lbPolicy` in there, though from the code I see that both are used independently.

**Comment by aattuluri** [comment]
@hzxuzhonghu Can you please do a final review? 
I spent some time on the failing tests, doesn't seem to be related to my feature.

File: tests/integration/pilot/locality/main_test.go [hunk 7]
```diff
@@ -125,7 +131,7 @@ func init() {
 func TestMain(m *testing.M) {
 	framework.NewSuite("locality_prioritized_failover_loadbalancing", m).
 		Label(label.CustomSetup).
-		SetupOnEnv(environment.Kube, istio.Setup(&ist, setupConfig)).
```
Comment (howardjohn) [inline tests/integration/pilot/locality/main_test.go]: In surprised lint doesn't fail here for setupConfig not being used anymore

**Review by howardjohn — APPROVED** [review]

**Review by hzxuzhonghu — APPROVED** [review]
Please fix the ut

**Comment by jasonwzm** [comment]
i think it is safe to get the latest api version by doing `go get istio.io/api@master` and after that you can do `make gen` so that the fix to the CRDs can be corrected. Plus the APIs have not changed, just some descriptions.

File: tests/integration/pilot/locality/main_test.go (hunk 7 above)
Comment (aattuluri) [reply tests/integration/pilot/locality/main_test.go]: It did after I fixed some other tests, weird, fixed it now.

**Comment by aattuluri** [comment]
@jasonwzm @hzxuzhonghu Need your help to figure out if these are flaky tests, the errors seem to be generic and not related to my change.

**Comment by howardjohn** [comment]
since they failed repeatedly on this PR but not others, I think its cause by this change.

That being said, I don't see why. May be some subtle issue or as a result of other changes pull in from api update?

**Comment by aattuluri** [comment]
> since they failed repeatedly on this PR but not others, I think its cause by this change.
> 
> That being said, I don't see why. May be some subtle issue or as a result of other changes pull in from api update?

Ok, probably I should try downgrading to some previous version of api on my branch and check?

**Comment by jasonwzm** [comment]
If you compare the API between your original change and the master, there is little changed on the API definitions themselves and the generated go types. https://github.com/istio/api/compare/9a5ec23e27ea566b0375d1516e2d0f8afe9567d8...master

**Comment by jasonwzm** [comment]
We added `MarshalJSON` methods to the APIs in the latest API. I am guessing that is messing around with our structpath implementation. I will look into that tomorrow. In the meantime, please try to use commit https://github.com/istio/api/commit/9fe6a7da36736eb708249d9a85071001eb11deb9. (gencheck probably will fail, but let's see if the other tests are okay.)

**Comment by jasonwzm** [comment]
@aattuluri the crd yaml in the api commit cannot be parsed by our test framework. that is why the tests are failing now. you can keep last crd-all.gen.yaml as unchanged.

**Comment by jasonwzm** [comment]
some values in our synthetic service entry tests are invalid. im trying to spot those and hopefully can fix the test failures we were seeing yesterday.

**Comment by aattuluri** [comment]
> @aattuluri the crd yaml in the api commit cannot be parsed by our test framework. that is why the tests are failing now. you can keep last crd-all.gen.yaml as unchanged.

Yep, reverted that.

**Comment by aattuluri** [comment]
> some values in our synthetic service entry tests are invalid. im trying to spot those and hopefully can fix the test failures we were seeing yesterday.

Thanks will great, thanks!

**Comment by jasonwzm** [comment]
Since I made commits to the PR, I will not be approving on behalf of config team. @ayj Can you please take a look? Particularly the fixes for the json naming in conformance tests and synthetic service entry integration tests, which I mentioned to you today. 

Unfortunately, this PR has grown to a huge one because the latest api commit had a few changes.

**Review by ayj — APPROVED** [review]

**Comment by jasonwzm** [comment]
@kyessenov Kuat, can you please review the changes of json naming in mixer tests?

**Review by kyessenov — APPROVED** [review]

**Comment by howardjohn** [comment]
@aattuluri thanks for implementing this and everything else that got dragged along with it!

**Comment by anilutcs** [comment]
@howardjohn Sure thing. Thanks for being proactive and helping on this one.

**Pull request merged**